talk/objectorientation/functors.tex

```diff
@@ -25,25 +25,28 @@
 \end{frame}
 
 \begin{frame}[fragile]
-  \frametitlecpp[98]{Function objects}
+  \frametitlecpp[20]{Function objects}
   \begin{block}{Typical usage}
     \begin{itemize}
-    \item pass a function to another one
-    \item or to an STL algorithm
+    \item pass a function to another one, like STL/ranges algorithm
     \end{itemize}
   \end{block}
+  \small
   \begin{cppcode}
-    struct BinaryFunction {
-      virtual double operator() (double a, double b) = 0;
-    };
-    double binary_op(double a, double b, BinaryFunction &func) {
-      return func(a, b);
+    int count_if(const auto& range, auto predicate) {
```

Comment:
To show a higher-order function here, we need templates. I chose to use `auto` here, since we already discussed the `auto` keyword and it should naturally follow what's going on here.

Reply:
The example is indeed better now, but it requires C++20 and makes an expert use of auto, which was not yet explained (and is not at all I believe). That was precisely the reason of being of the original base class !

To word it differently, you're replacing inheritance by templating. It's almost always a good move, expect we did not yet introduce templating at that stage.

Not sure what the conclusion is. @hageboeck any opinion ?

Reply:
Hmm, I agree that it goes two steps ahead. Maybe we should point out that it's a template, and that these are going to be explained later?

Reply:
Thinking about it, I don't see how to do it without templates or inheritance. Maybe we should indeed show "the better way" of doing it, give a brief explanation (verbally) what's going on, and refer to the template chapter on the slide.

Reply:
Ok, let's go for that. Last detail, @bernhardmgruber can you add a godbolt URL ?